### PR TITLE
Dao refactoring

### PR DESCRIPTION
**Pull request by Swaglis**
refactored dao layer so we dont have to repeat basic methods in every dao object implementation

### CODE REVIEW

File: formulaoneteam-persistence/src/main/java/cz/muni/fi/pa165/dao/base/DaoImpl.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
+/*
+ * To change this license header, choose License Headers in Project Properties.
+ * To change this template file, choose Tools | Templates
+ * and open the template in the editor.
+ */
```
Comment (elderanakain) [inline formulaoneteam-persistence/src/main/java/cz/muni/fi/pa165/dao/base/DaoImpl.java]: Please, remove it from files and change settings not to add this message automatically.

File: formulaoneteam-persistence/src/main/java/cz/muni/fi/pa165/dao/base/DaoImpl.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+import java.lang.reflect.ParameterizedType;
+import java.util.List;
+import javax.persistence.EntityManager;
+import javax.persistence.PersistenceContext;
+import org.springframework.lang.Nullable;
+import org.springframework.stereotype.Repository;
+
+/**
+ *
+ * @author G
```
Comment (elderanakain) [inline formulaoneteam-persistence/src/main/java/cz/muni/fi/pa165/dao/base/DaoImpl.java]: Please, fix the `javadoc`.

File: formulaoneteam-persistence/src/main/java/cz/muni/fi/pa165/dao/base/DaoUser.java [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+import javax.persistence.NoResultException;
+import org.springframework.lang.Nullable;
+
+import static cz.muni.fi.pa165.entity.base.User.EMAIL_FIELD;
+import javax.persistence.TypedQuery;
+
+/**
+ *
+ * @author G
+ * @param <T>
```
Comment (elderanakain) [inline formulaoneteam-persistence/src/main/java/cz/muni/fi/pa165/dao/base/DaoUser.java]: Please, fix the `javadoc`.

File: formulaoneteam-persistence/src/main/java/cz/muni/fi/pa165/entity/Engineer.java [hunk 4]
```diff
@@ -0,0 +1,24 @@
+package cz.muni.fi.pa165.entity;
+
+import cz.muni.fi.pa165.entity.base.User;
+import javax.persistence.Column;
+import javax.validation.constraints.NotNull;
+
+/**
+ *
+ * @author G
```
Comment (elderanakain) [inline formulaoneteam-persistence/src/main/java/cz/muni/fi/pa165/entity/Engineer.java]: Please, fix the `javadoc`.

File: formulaoneteam-persistence/src/test/java/cz/muni/fi/pa165/entity/base/UserTest.java [hunk 5]
```diff
@@ -15,10 +15,10 @@
 
     @BeforeMethod
     public void setUp() {
-        user = new User() {};
+        user = new User() {
+        };
```
Comment (elderanakain) [inline formulaoneteam-persistence/src/test/java/cz/muni/fi/pa165/entity/base/UserTest.java]: Please, revert this change. Don't need to have it this way.

File: formulaoneteam-persistence/src/test/java/cz/muni/fi/pa165/entity/base/BaseEntityTest.java [hunk 6]
```diff
@@ -14,7 +14,8 @@
 
     @BeforeMethod
     public void setUp() {
-        baseEntity = new BaseEntity() {};
+        baseEntity = new BaseEntity() {
```
Comment (elderanakain) [inline formulaoneteam-persistence/src/test/java/cz/muni/fi/pa165/entity/base/BaseEntityTest.java]: Please, revert this change. Don't need to have it this way.

File: formulaoneteam-persistence/src/test/java/cz/muni/fi/pa165/dao/manager/ManagerDaoTest.java [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+    @Test
+    public void findByEmail() {
+        //given
+        String email = "asd@asd.asd";
+        Manager manager = new Manager();
+        manager.setName("asd");
+        manager.setPassword("asd");
+        manager.setSurname("asd");
+        manager.setEmail(email);
+        // when(manager.isConfigured()).thenReturn(true);
```
Comment (elderanakain) [inline formulaoneteam-persistence/src/test/java/cz/muni/fi/pa165/dao/manager/ManagerDaoTest.java]: Why do you have commented code here?

File: formulaoneteam-persistence/src/test/java/cz/muni/fi/pa165/dao/manager/ManagerDaoTest.java [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+
+        //when
+        managerDao.add(manager);
+
+        AssertJUnit.assertNotNull(managerDao.findByEmail(email));
```
Comment (elderanakain) [inline formulaoneteam-persistence/src/test/java/cz/muni/fi/pa165/dao/manager/ManagerDaoTest.java]: It is better to use static import for `AssertJUnit` in order to avoid it repeating for each assert call.

File: formulaoneteam-persistence/src/test/java/cz/muni/fi/pa165/dao/manager/ManagerDaoTest.java [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+
+        //when
+        managerDao.add(manager);
+
+        //then
+        AssertJUnit.assertNotNull(managerDao.findById(manager.getId()));
+    }
+
+    @Test
+    public void findByEmail() {
```
Comment (elderanakain) [inline formulaoneteam-persistence/src/test/java/cz/muni/fi/pa165/dao/manager/ManagerDaoTest.java]: Please, rename this test to something more meaningful like - `whatYouTest_expectedResult`. Because one test for this method is not enough, we need to test when something is broken.

File: formulaoneteam-persistence/src/main/java/cz/muni/fi/pa165/entity/base/User.java [hunk 10]
```diff
@@ -87,14 +87,17 @@ public boolean hasPassword() {
...
     @Override
     public boolean equals(Object o) {
-        if (this == o) return true;
-        if (!(o instanceof User)) return false;
+        if (this == o) {
+            return true;
+        }
+        if (!(o instanceof User)) {
+            return false;
+        }
```
Comment (elderanakain) [inline formulaoneteam-persistence/src/main/java/cz/muni/fi/pa165/entity/base/User.java]: I guess no need for brackets here. I know that it is better to have them for if clauses, but it is okay to avoid them for `equals` methods because of the repeating pattern.

File: formulaoneteam-persistence/src/main/java/cz/muni/fi/pa165/entity/EngineerSpecialization.java [hunk 11]
```diff
@@ -0,0 +1,10 @@
+package cz.muni.fi.pa165.entity;
+
+/**
+ *
+ * @author G
+ */
```
Comment (elderanakain) [inline formulaoneteam-persistence/src/main/java/cz/muni/fi/pa165/entity/EngineerSpecialization.java]: Please, fix the `javadoc`.

File: formulaoneteam-persistence/src/main/java/cz/muni/fi/pa165/entity/EngineerSpecialization.java [hunk 12]
```diff
@@ -0,0 +1,10 @@
...
+public enum EngineerSpecialization {
+    ENGINES,
+    AERODYNAMICS
+}
```
Comment (elderanakain) [inline formulaoneteam-persistence/src/main/java/cz/muni/fi/pa165/entity/EngineerSpecialization.java]: Why do you have only these specializations here?

File: formulaoneteam-persistence/src/main/java/cz/muni/fi/pa165/entity/Engineer.java [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+ */
+public class Engineer extends User {
```
Comment (elderanakain) [inline formulaoneteam-persistence/src/main/java/cz/muni/fi/pa165/entity/Engineer.java]: Please, add `@Entity` annotation here.

File: formulaoneteam-persistence/pom.xml [hunk 14]
```diff
@@ -91,6 +91,12 @@
             <version>3.0</version>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>javax.inject</groupId>
+            <artifactId>javax.inject</artifactId>
+            <version>1</version>
+            <scope>test</scope>
+        </dependency>
```
Comment (elderanakain) [inline formulaoneteam-persistence/pom.xml]: I don't quite get it why there are these changes because I have already added Inject for test scope.

File: formulaoneteam-persistence/src/main/java/cz/muni/fi/pa165/dao/base/DaoImpl.java [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+ * @param <T>
+ */
+@Repository
```
Comment (elderanakain) [inline formulaoneteam-persistence/src/main/java/cz/muni/fi/pa165/dao/base/DaoImpl.java]: I think that it is better to mark only final classes as repositories. Abstract classes cannot be solo repositories.

File: formulaoneteam-persistence/src/main/java/cz/muni/fi/pa165/dao/base/DaoImpl.java [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+    @PersistenceContext
+    protected EntityManager entityManager;
+
+    public DaoImpl(Class<T> classType) {
+        this.classType = classType;
+    }
+
+    @Override
+    public T findById(Long id) {
+        return (T)entityManager.find(classType, id);
```
Comment (elderanakain) [inline formulaoneteam-persistence/src/main/java/cz/muni/fi/pa165/dao/base/DaoImpl.java]: The cast is redundant.

File: formulaoneteam-persistence/src/main/java/cz/muni/fi/pa165/dao/base/DaoImpl.java [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+    public void update(@Nullable T entity) {
+        validateEntity(entity);
+
+        entityManager.merge(entity);
+    }
+
+    @Override
+    public List<T> findAll() {
+        return entityManager
+                .createQuery("FROM " + classType.getName(), classType)
```
Comment (elderanakain) [inline formulaoneteam-persistence/src/main/java/cz/muni/fi/pa165/dao/base/DaoImpl.java]: The query doesn't seem right. I would say that `SELECT` part is missing.

File: formulaoneteam-persistence/src/main/java/cz/muni/fi/pa165/dao/base/DaoUser.java [hunk 18]
```diff
@@ -0,0 +1,23 @@
...
+ */
+public interface DaoUser<T extends User> extends Dao<T> {
```
Comment (elderanakain) [inline formulaoneteam-persistence/src/main/java/cz/muni/fi/pa165/dao/base/DaoUser.java]: Please, rename to `UserDao`.

File: formulaoneteam-persistence/src/main/java/cz/muni/fi/pa165/dao/base/DaoUserImpl.java [hunk 19]
```diff
@@ -0,0 +1,31 @@
...
+
+import cz.muni.fi.pa165.entity.base.User;
+import static cz.muni.fi.pa165.entity.base.User.EMAIL_FIELD;
+import javax.persistence.NoResultException;
+
+/**
+ *
+ * @author Ivan Dendis
+ */
+public abstract class DaoUserImpl<T extends User> extends DaoImpl<T>
```
Comment (elderanakain) [inline formulaoneteam-persistence/src/main/java/cz/muni/fi/pa165/dao/base/DaoUserImpl.java]: Please, rename to `UserDaoImpl`.

File: formulaoneteam-persistence/src/main/java/cz/muni/fi/pa165/dao/base/DaoImpl.java [hunk 20]
```diff
@@ -0,0 +1,67 @@
...
+@Repository
+public abstract class DaoImpl<T extends BaseEntity> implements Dao<T> {
+    protected final Class<T> classType;
+
+    @PersistenceContext
+    protected EntityManager entityManager;
+
+    public DaoImpl(Class<T> classType) {
+        this.classType = classType;
+    }
```
Comment (elderanakain) [inline formulaoneteam-persistence/src/main/java/cz/muni/fi/pa165/dao/base/DaoImpl.java]: It is better to have `protected abstract` method as for `validateEntity`. We want to have static `classType` for each final Dao implementation. No need to pass it via the constructor.

File: formulaoneteam-persistence/src/main/java/cz/muni/fi/pa165/dao/base/DaoUserImpl.java [hunk 21]
```diff
@@ -0,0 +1,31 @@
...
+        implements DaoUser<T> {
+
+    public DaoUserImpl(Class<T> classType) {
+        super(classType);
+    }
+
+    @Override
+    public T findByEmail(String email) {
+        try {
+            String queryString = "FROM " + classType.getName()
```
Comment (elderanakain) [inline formulaoneteam-persistence/src/main/java/cz/muni/fi/pa165/dao/base/DaoUserImpl.java]: Please, fix the query string.

File: formulaoneteam-persistence/src/main/java/cz/muni/fi/pa165/dao/engineer/EngineerDao.java [hunk 22]
```diff
@@ -0,0 +1,11 @@
+package cz.muni.fi.pa165.dao.engineer;
+
+import cz.muni.fi.pa165.dao.base.DaoUser;
+import cz.muni.fi.pa165.entity.Engineer;
+
+/**
+ *
+ * @author G
+ */
```
Comment (elderanakain) [inline formulaoneteam-persistence/src/main/java/cz/muni/fi/pa165/dao/engineer/EngineerDao.java]: Please, fix the `javadoc`.

File: formulaoneteam-persistence/pom.xml (hunk 14 above)
Comment (Swaglis) [reply formulaoneteam-persistence/pom.xml]: i cherry picked your commit from develop, maybe thats why

File: formulaoneteam-persistence/src/main/java/cz/muni/fi/pa165/dao/base/DaoImpl.java (hunk 17 above)
Comment (elderanakain) [reply formulaoneteam-persistence/src/main/java/cz/muni/fi/pa165/dao/base/DaoImpl.java]: Hah, that's true. Didn't know that 👍

File: formulaoneteam-persistence/src/test/java/cz/muni/fi/pa165/dao/base/DaoTest.java [hunk 23]
```diff
@@ -0,0 +1,77 @@
...
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.verify;
+import static org.testng.Assert.assertTrue;
+
+/**
+ * Tests for the general Dao functionality.
+ *
+ * @author Ivan Dendis
+ */
+public class DaoTest {
```
Comment (elderanakain) [inline formulaoneteam-persistence/src/test/java/cz/muni/fi/pa165/dao/base/DaoTest.java]: Please, do not try to create a generic test for `Dao`. Each of us should create unit tests for someone's Dao. Therefore, it would complicate things and, I guess, wouldn't be accepted by our supervisor.
